### ISSUE TITLE
restrict group by to only group by fields with < 30 fields

### ISSUE THREAD

**Issue by prabhasp** [SEL-Columbia/bamboo#128]
creating an issue to implement one of the lessons of #127, and assigning to @myf for now. @pld @mejymejy feel free to close / change assignments / milestone as you see appropriate.

**Comment by pld** [comment]
An alternative is to return, in the dataset info, the number of unique entries per column, then the client can't choose which columns to allow group by on based on this information.

**Comment by prabhasp** [comment]
@pld that would make the dataset info dynamic (rather than static). do we want that? since this isn't decided, i'm moving to 0.4, but we should decide soon.

**Comment by pld** [comment]
good point, perhaps not.  But, if the client doesn't have this info they can't tell what groupings are allowed without sending a request and seeing if it is denied or counting uniques themselve.

If this restriction exists they would like to know what groupings are allowed in order to determine what to show in a UI.

**Comment by prabhasp** [comment]
On Fri, Sep 7, 2012 at 2:44 PM, Peter Lubell-Doughtie <
notifications@github.com> wrote:

> good point, perhaps not. But, if the client doesn't have this info they
> can't tell what groupings are allowed without sending a request and seeing
> if it is denied or counting uniques themselve.
> 
> If this restriction exists they would like to know what groupings are
> allowed in order to determine what to show in a UI.
> 
> Therefore my first proposal that only those string fields with < 30
> groupings should be (automatically) labeled with  "olap_type" "dimension"
> by bamboo.

Note that a key_count on a summary query _will_ give this information to
the client, although in a bit of a convoluted way.

>  —
> Reply to this email directly or view it on GitHubhttps://github.com/modilabs/bamboo/issues/128#issuecomment-8374702.

## 

Prabhas Pokharel
http://twitter.com/prabhasp
US mobile: +1 347 948 7654
skype/facebook/whatever: prabhasp

**Comment by pld** [comment]
But that would also be dynamic, the number of groupings changes as the data changes.

**Comment by prabhasp** [comment]
fair point. arbitrary restriction (such as < 30 unique fields) paired with
something to find out what groupbys are allowed? or just leave it to the
client still?

On Fri, Sep 7, 2012 at 2:59 PM, Peter Lubell-Doughtie <
notifications@github.com> wrote:

> But that would also be dynamic, the number of groupings changes as the
> data changes.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/modilabs/bamboo/issues/128#issuecomment-8375195.

## 

Prabhas Pokharel
http://twitter.com/prabhasp
US mobile: +1 347 948 7654
skype/facebook/whatever: prabhasp

**Comment by pld** [comment]
Mark and I we're talking about this a bit ago, we notice that the number of unique fields in a column is available to the client by summarizing and counting the keys for that column (since number bins is equivalent to number of unique columns).  This feels like a question for those building clients: do you want to see the number of unique values for each column in another way?

**Comment by prabhasp** [comment]
Peter, I think for me, this is about whether we should be limiting
expensive performance to clients. If you allow expensive computations,
clients will perfrom them, and the main impact is on the server / dev ops
side of things. Should we be limiting expensive computations, or at least
making expensive computations more explicit / harder to perform?
Personally, I think a reasonable strategy is following Cass Sunstein et.
al's "nudge" philosophy, making less expensive computations easy, and
requiring extra specification for the more expensive cases.

However, after discussions later in the week, IMHO the right approach to
this particular expensive operation (group by) is #136, and therefore, this
issue can be closed in favor of that other one.

On Friday, September 14, 2012, Peter Lubell-Doughtie wrote:

> Mark and I we're talking about this a bit ago, we notice that the number
> of unique fields in a column is available to the client by summarizing and
> counting the keys for that column (since number bins is equivalent to
> number of unique columns). This feels like a question for those building
> clients: do you want to see the number of unique values for each column in
> another way?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/modilabs/bamboo/issues/128#issuecomment-8561828.

## 

Prabhas Pokharel
http://twitter.com/prabhasp
US mobile: +1 347 948 7654
skype/facebook/whatever: prabhasp

**Comment by prabhasp** [comment]
proposal: close, given #136

**Comment by myf** [comment]
I think the schema should give an idea of 1) what are the fields 2) what are the type of the fields 3) how computational intensive if one were to group by these fields. With the unique_value it is definitely welcomed on the client side. There re asome technical limitation though. Currently the schema is generated right away before the formation of the summary record, while in order to fetch the unique_values bamboo would have to go in summary. We'll see and hopefully get it through soon.

**Comment by mejymejy** [comment]
first of all i think that it makes sense to have something that tells you about the unique values that is core to the api (you shouldn't have to take the length of keys to an array).  that being said i think there are a least a couple of ways to look at this...
1. relevant information about the dataset should be in the info api call -- this means that somewhere (probably inside the dataset schema) there would be listed all of the info that you need to characterize the dataset including dimension/measure and "uniqueness".  i think that this is nice because then the api user can make one call to get the info about the dataset and choose to make further calls based on that call alone
2. the uniqueness is a property of the column that should be included in the summary api call -- this would treat the information about the column as a sort of a preliminary analysis of the column data like the other stuff in there (discrete value counts, basic stats, etc.) and you would need to make a specific api call to obtain this info.  this is how it currently is, though not explicitly since you need to take the length of the array for value counts to get the unique values for a column.
3. do both -- DENORMALIZE FTW

note: the schema is already dynamic since we are counting the number of columns/rows and including them.  this is not an issue, however, since whenever we change the dataset we update the schema.

one of the questions that i have a hard time answering is this:  should the schema be dynamic (semantically)?  my intuition says that it should only include information about the dataset that isn't changing, but when i really think about it i tend to believe that it should really just include as much relevant info as possible to make life easier for the api user...

tl;dr i can't make decisions
